pkg/controller/association/reconciler.go

```diff
@@ -294,10 +296,11 @@ func (r *Reconciler) reconcileAssociation(ctx context.Context, association commo
 
 	// construct the expected association configuration
 	expectedAssocConf := &commonv1.AssociationConf{
-		CACertProvided: caSecret.CACertProvided,
-		CASecretName:   caSecret.Name,
-		URL:            url,
-		Version:        ver,
+		CACertProvided:        caSecret.CACertProvided,
+		CASecretName:          caSecret.Name,
+		AdditionalSecretsHash: fmt.Sprint(secretsHash.Sum32()),
```

Comment:
Actually I think we should always include the hash because we have a bug with the CA rotation with Agent even if Agent/FleetServer and ES are in the same namespace:
```diff
diff --git a/pkg/controller/association/controller/agent_fleetserver.go b/pkg/controller/association/controller/agent_fleetserver.go
index 7a3ebc4a9..89bda5264 100644
--- a/pkg/controller/association/controller/agent_fleetserver.go
+++ b/pkg/controller/association/controller/agent_fleetserver.go
@@ -74,12 +74,6 @@ func additionalSecrets(ctx context.Context, c k8s.Client, assoc commonv1.Associa
 		return nil, err
 	}
 
-	// if both agent and ES are same namespace no copying needed
-	if agent.GetNamespace() == esAssociation.GetNamespace() {
-		log.V(1).Info("no additional secrets because same namespace")
-		return nil, nil
-	}
-
 	conf, err := esAssociation.AssociationConf()
 	if err != nil {
 		log.V(1).Info("no additional secrets because no assoc conf")
diff --git a/pkg/controller/association/reconciler.go b/pkg/controller/association/reconciler.go
index d550d0c3a..a552873d2 100644
--- a/pkg/controller/association/reconciler.go
+++ b/pkg/controller/association/reconciler.go
@@ -271,7 +271,7 @@ func (r *Reconciler) reconcileAssociation(ctx context.Context, association commo
 			return commonv1.AssociationPending, err // maybe not created yet
 		}
 		for _, sec := range additionalSecrets {
-			if err := copySecret(ctx, r.Client, secretsHash, association.GetNamespace(), sec); err != nil {
+			if err := copySecret(ctx, r.Client, secretsHash, k8s.ExtractNamespacedName(association.Associated()), association.GetNamespace(), sec); err != nil {
 				return commonv1.AssociationPending, err
 			}
 		}
diff --git a/pkg/controller/association/secret.go b/pkg/controller/association/secret.go
index e8e746767..d0569c63b 100644
--- a/pkg/controller/association/secret.go
+++ b/pkg/controller/association/secret.go
@@ -195,14 +195,18 @@ func filterManagedElasticRef(associations []commonv1.Association) []commonv1.Ass
 }
 
 // copySecret will copy the source secret to the target namespace adding labels from the associated object to ensure garbage collection happens.
-func copySecret(ctx context.Context, client k8s.Client, secHash hash.Hash, targetNamespace string, source types.NamespacedName) error {
+func copySecret(ctx context.Context, client k8s.Client, secHash hash.Hash, associated types.NamespacedName, targetNamespace string, source types.NamespacedName) error {
 	var original corev1.Secret
 	if err := client.Get(ctx, source, &original); err != nil {
 		return err
 	}
+	commonhash.WriteHashObject(secHash, original.Data)
+	if targetNamespace == original.Namespace {
+		return nil
+	}
+
 	var expected corev1.Secret
 	expected.Data = original.Data
-	commonhash.WriteHashObject(secHash, original.Data)
 	expected.Labels = original.Labels
 	expected.Annotations = original.Annotations
 	expected.Name = original.Name
```

This is because we do not trigger a reconciliation in the Agent controller if the ES CA changes. Always including the hash annotation would make sure that the reconciliation happens (thanks to the annotation update) and the existing code takes then care of rotating the Agent pods because it looks at the secret contents of the ES CA secret.

Reply:
> Should we include the hash only if AdditionalSecrets is not empty?

I think you are right that we can omit the hash when there are not additional secrets. My comment was more aiming at the case where there are additional secrets but they are in the same namespace as the Agent.